docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,235 @@
+
+<details><summary>List of illegal constructs in type functions</summary>
+
+* `while` / `repeat` loops
+* invoking other type functions / regular functions / lambdas
+    * we will not (and probably never) allow type functions to call regular functions for the sake of sandboxing the runtime and analysis.
+* referring to locals / globals in the outer scope
+* global functions: `getfenv`, `setfenv`, `pcall`, `xpcall`, `require`
+* libraries: `coroutine`, `debug`, `string.gsub`
+
+Note: we are aware that for loops can cause infinite runtime. For the time being, we will not be handling this case. In the event that a developer accidentally creates an infinitely long type function, they will need to fix the type function and restart the analysis.
+
+</details>
+
+To give warnings, developers can use `print()` with custom warning messages. To fail reductions, developers can use `error()` with custom error messages. If nothing is returned by the type function, it will fail to reduce with the default message: "Failed to reduce \<Name\> type function with no return values".
+
+To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `typelib`. A `typelib` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, the `typelib` library will closely mimic the implementation of static types in Luau Analysis. Most importantly, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 
+
+<details><summary>typelib library methods (dropdown)</summary>
+
+Methods under a different type heading (ex: `Singleton`) imply that the methods are only available for those types. At the implementation level, there is a check to make sure that the type-specific methods are being called on the correct types. For instance, `getindexer()` asserts that `istable()` is true.
+
+#### typelib
+All attributes of newly created typelib are initialized with empty tables / arrays and `typelib:getnil()`. For instance, `typelib:newtable()` initializes its properties with an empty table and index / index result type as `typelib:getnil()`.
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getnil()` | `typelib` | returns an immutable runtime representation of the built-in type `nil` |
+| `getunknown()` | `typelib` | returns an immutable runtime representation of the built-in type `unknown` |
+| `getnever()` | `typelib` | returns an immutable runtime representation of the built-in type `never` |
+| `getany()` | `typelib` | returns an immutable runtime representation of the built-in type `any` |
+| `getnegation(arg: typelib)` | `typelib` | returns an immutable runtime representation of the negation of the argument; the argument cannot be `istable()`, `ismetatable` or `isfunction()` |
```

Comment:
The presence of this basically means we have to do the negation syntax RFC (since otherwise it means people will just write trivial type functions for negations). This is okay, of course, since that was already planned, but just a thing to note.